libs/gltfio/src/ReusableAnimator.cpp

```diff
@@ -0,0 +1,489 @@
+    switch (valuesAccessor->type) {
+        case cgltf_type_scalar:
+            dst.values.resize(valuesAccessor->count);
+            cgltf_accessor_unpack_floats(src.output, &dst.values[0], valuesAccessor->count);
+            break;
+        case cgltf_type_vec3:
+            dst.values.resize(valuesAccessor->count * 3);
+            cgltf_accessor_unpack_floats(src.output, &dst.values[0], valuesAccessor->count * 3);
+            break;
+        case cgltf_type_vec4:
+            dst.values.resize(valuesAccessor->count * 4);
+            cgltf_accessor_unpack_floats(src.output, &dst.values[0], valuesAccessor->count * 4);
+            break;
+        default:
+            GLTFIO_WARN("Unknown animation type.");
+            return;
+    }
+
+    switch (src.interpolation) {
+        case cgltf_interpolation_type_linear:
+            dst.interpolation = Sampler::LINEAR;
+            break;
+        case cgltf_interpolation_type_step:
+            dst.interpolation = Sampler::STEP;
+            break;
+        case cgltf_interpolation_type_cubic_spline:
+            dst.interpolation = Sampler::CUBIC;
+            break;
+    }
+}
+
+static void setTransformType(const cgltf_animation_channel& src, Channel& dst) {
+    switch (src.target_path) {
+        case cgltf_animation_path_type_translation:
+            dst.transformType = Channel::TRANSLATION;
+            break;
+        case cgltf_animation_path_type_rotation:
+            dst.transformType = Channel::ROTATION;
+            break;
+        case cgltf_animation_path_type_scale:
+            dst.transformType = Channel::SCALE;
+            break;
+        case cgltf_animation_path_type_weights:
+            dst.transformType = Channel::WEIGHTS;
+            break;
+        case cgltf_animation_path_type_invalid:
+            GLTFIO_WARN("Unsupported channel path.");
+            break;
+    }
+}
+
+static bool validateAnimation(const cgltf_animation& anim) {
+    for (cgltf_size j = 0; j < anim.channels_count; ++j) {
+        const cgltf_animation_channel& channel = anim.channels[j];
+        const cgltf_animation_sampler* sampler = channel.sampler;
+        if (!channel.target_node) {
+            continue;
+        }
+        if (!channel.sampler) {
+            return false;
+        }
+        cgltf_size components = 1;
+        if (channel.target_path == cgltf_animation_path_type_weights) {
+            if (!channel.target_node->mesh || !channel.target_node->mesh->primitives_count) {
+                return false;
+            }
+            components = channel.target_node->mesh->primitives[0].targets_count;
+        }
+        cgltf_size values = sampler->interpolation == cgltf_interpolation_type_cubic_spline ? 3 : 1;
+        if (sampler->input->count * components * values != sampler->output->count) {
+            return false;
+        }
+    }
+    return true;
+}
+
+ReusableAnimator::ReusableAnimator(FilamentAsset* animationAsset) {
```

Comment:
On creation, we create the samplers but not the channels. So we have all the animation data, but they are not being applied to any entities yet. We add the channels in `addAnimatedAsset`